fundraiser/src/main.sw

```diff
@@ -0,0 +1,502 @@
+
+        // The author can only claim the pledges once the target amount has been reached otherwise
+        // users should be able to withdraw
+        require(campaign_info.target_amount <= campaign_info.total_pledge, CampaignError::TargetNotReached);
+
+        // The author can only claim once to prevent the entire contract from being drained
+        require(!campaign_info.claimed, UserError::AlreadyClaimed);
+
+        // The author cannot claim after they have cancelled the campaign regardless of any other
+        // checks
+        require(!campaign_info.cancelled, CampaignError::CampaignHasBeenCancelled);
+
+        // Mark the campaign as claimed and overwrite the previous state with the updated version
+        campaign_info.claimed = true;
+        storage.campaign_info.insert(id, campaign_info);
+
+        // Transfer the total pledged to this campaign to the beneficiary
+        transfer(campaign_info.total_pledge, campaign_info.asset, campaign_info.beneficiary);
+
+        // We have updated the state of a campaign therefore we must log it
+        log(ClaimedEvent {
+            id
+        });
+    }
+
+    /// Allows a user to pledge any amount of the campaign asset towards the campaign goal
+    ///
+    /// In order to reach the campaign's target amount users must pledge some amount of asset towards
+    /// that campaign.
+    /// This information is recorded for the campaign and for the user so that they can unpledge.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique campaign identifier which is a number from the storage.total_campaigns range
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created
+    /// * When the user attempts to pledge when the deadline has been reached
+    /// * When the user pledges a different asset to the one specified in the campaign
+    /// * When the user pledges after the campaign has been cancelled
+    /// * When an AuthError is generated
+    #[storage(read, write)]fn pledge(id: u64) {
+        // User cannot interact with a non-existent campaign
+        validate_id(id, storage.total_campaigns);
+
+        // Retrieve the campaign in order to check its data / update it
+        let mut campaign_info = storage.campaign_info.get(id);
+
+        // The users should only have the ability to pledge to campaigns that have not reached their
+        // deadline (ended naturally - not been cancelled)
+        require(height() < campaign_info.deadline, CampaignError::CampaignEnded);
+
+        // The campaign specifies an asset that it accepts therefore the user must pledge the correct
+        // asset in order to update the state of the campaign
+        require(campaign_info.asset == msg_asset_id(), UserError::IncorrectAssetSent);
+
+        // A user cannot pledge zero since it does not make sense to do so
+        require(0 < msg_amount(), UserError::AmountCannotBeZero);
```

Comment:
Maybe check this even before reading the campaign_info from storage. If `msg_amount() == 0` there's no point doing a storage read.

Reply:
I think the order makes sense as it is.
When you pledge you must check that what you are pledging to is active, what you are trying to pledge is the correct type and the amount is also correct.
Trying to change the order doesn't make sense in the real world.

If and when we are able to extract the error messages then it will become more relevant.

Reply:
I disagree. 
Of all the checks needed in this function, this is the only one that can actually be done before reading from storage (and incurring extra gas costs).
No point in reading storage and making a number of other checks first if we already know the function is going to revert based on `msg_value()`.

In general, "Fail hard and fail early" :)

Reply:
This is not something that I will budge on. Semantically it makes sense to perform the checks this way and if you're trying to break the system then you deserve to be punished. This only costs those who intentially pledge 0. No one who has any brain cells will actually pledge 0 because that doesn't make sense to do so therefore if you are trying to perform boundary conditions / hacks then that additional cost is on you.

That being said I do agree with your last sentence.